llvm/lib/Target/AIE/AIEInterBlockScheduling.cpp

```diff
@@ -343,54 +412,94 @@ MachineInstr *InterBlockScheduling::latencyConverged(BlockState &BS) const {
   return nullptr;
 }
 
-bool InterBlockScheduling::updateFixPoint(BlockState &BS) {
+SchedulingStage InterBlockScheduling::updateFixPoint(BlockState &BS) {
+  if (BS.Kind != BlockType::Loop) {
+    return BS.FixPoint.Stage = SchedulingStage::SchedulingDone;
+  }
+
   if (!BS.FixPoint.NumIters) {
     // This is the first time we have scheduled this loop. In that first
     // iteration, we have recorded the region decomposition.
     // Now we can create the interblock edges between the top and the bottom
     // region
-    BS.initInterBlock(*Context);
+    BS.initInterBlock(*Context, *HR);
   }
 
-  if (MachineInstr *MINeedsHigherCap = latencyConverged(BS)) {
-    auto Res = BS.FixPoint.PerMILatencyMargin.try_emplace(MINeedsHigherCap, 0);
-    // Increase the latency margin per instruction, unless we already iterated
-    // more than MaxExpensiveIterations without converging.
-    if (BS.FixPoint.NumIters <= MaxExpensiveIterations) {
-      ++Res.first->second;
-    } else {
-      BS.FixPoint.LatencyMargin++;
+  BS.FixPoint.NumIters++;
+  int &II = BS.FixPoint.II;
+  if (!II) {
```

Comment:
That's now a big function with lots of `if`s. How about introducing two functions for checking the fixpoint state? E.g.

```
BS.FixPoint.NumIters++;

if (State== Scheduling)
  return checkLoopAwareScheduling();
return checkPipelining()
```

Reply:
(That would also help limit the noise in diff lines)